part-1.py

```diff
@@ -4,17 +4,44 @@
 # the appropriate comment.
 
 # factorial
-
+def factorial(n):
+    if n==0:
+        return 1
+    if n < 0:
+        raise ValueError
+    else:
+        return n * factorial(n - 1)
 
 
 # reverse
-
+def reverse(text):
+    if len(text) == 0:
+        return text
+    return reverse(text[1:]) + text[0]
 
 
 # bunny
-
+def bunny(count):
+    if count == 0:
+        return 0
+        
+    return 2 + bunny(count - 1)
 
 
 # is_nested_parens
-
+def is_nested_parens(parens, left_index = 0):
```

Comment:
This is very clever! Not necessarily better or worse than changing the `parens` string through the recursive calls, but not the typical way to solve it, which I appreciate seeing!